services/CurrentTime/include/ble-service-current-time/CurrentTimeService.h

```diff
@@ -0,0 +1,178 @@
+        uint8_t  hours;
+        uint8_t  minutes;
+        uint8_t  seconds;
+
+        uint8_t  weekday;
+
+        uint8_t  fractions256;
+
+        uint8_t  adjustReason;
+    };
+
+    struct EventHandler {
+        /**
+         * On current time changed
+         *
+         * This function is called if the current time characteristic is changed by the client
+         *
+         * @attention This is an abstract function and should be overridden by the user.
+         */
+        virtual void on_current_time_written(time_t current_time) {}
+    };
+
+    /**
+     * Set event handler
+     *
+     * @param handler EventHandler object to handle events raised by the current time service
+     */
+    void set_event_handler(EventHandler *handler) {
+
+    }
+
+    /**
+     * Construct and initialize a current time service.
+     *
+     * @param ble
+     * @param event_queue
+     * @param chainable_gap_event_handler
+     */
+    CurrentTimeService(BLE &ble, events::EventQueue &event_queue) :
+            _ble(ble),
+            _event_queue(event_queue),
+            _current_time_char(GattCharacteristic::UUID_CURRENT_TIME_CHAR, &_current_time) {
+        memset(&_current_time, 0, sizeof(_current_time));
+    }
+
+    ble_error_t init() {
+        GattCharacteristic *charTable[] = {&_current_time_char};
+        GattService currentTimeService(GattService::UUID_CURRENT_TIME_SERVICE, charTable, 1);
+
+        _current_time_char.setReadAuthorizationCallback(this, &CurrentTimeService::onDataRead);
+        _current_time_char.setReadAuthorizationCallback(this, &CurrentTimeService::onDataRead);
+
+        ble_error_t bleError = _ble.gattServer().addService(currentTimeService);
+
+        return bleError;
+    }
+
+    time_t get_current_time() {
+        time_t t = time(NULL);
+
+        return _current_time_offset + t;
+    }
+
+    void set_time(time_t new_time) {
+        time_t t = time(NULL);
+
+        _current_time_offset = t - new_time;
+    }
+
+private:
+    void set_current_time() {
+        timeval tv{};
+        gettimeofday(&tv, NULL);
+        // printf("%u\r\n", (unsigned int) tv.tv_sec);
+
+        struct tm *current_time;
+        time_t local_time = tv.tv_sec + _current_time_offset;
+        current_time = localtime(&local_time);
+
+        _current_time.year    = current_time->tm_year + 1900;
+        _current_time.month   = current_time->tm_mon  + 1;
+        _current_time.day     = current_time->tm_mday;
+        _current_time.hours   = current_time->tm_hour;
+        _current_time.minutes = current_time->tm_min;
+        _current_time.seconds = current_time->tm_sec;
+        _current_time.weekday = (current_time->tm_wday == 0) ? 7 : current_time->tm_wday;
+        // 1/256th of seconds not supported
+        _current_time.fractions256 = 0;
+    }
+
+    void onDataRead(GattReadAuthCallbackParams *read_request) {
+        set_current_time();
+
+        _ble.gattServer().write(_current_time_char.getValueHandle(),
```

Comment:
I'm not sure it is necessary to write the characteristic. The request expose a `data` pointer and `len` field that can be changed to point to `_current_time`.